python/sglang/srt/managers/schedule_batch.py

```diff
@@ -0,0 +1,909 @@
+        )
+        self.top_ks = torch.tensor(
+            [r.sampling_params.top_k for r in reqs], dtype=torch.int, device=device
+        )
+
+        # Each penalizers will do nothing if they evaluate themselves as not required by looking at
+        # the sampling_params of the requests (See {_is_required()} of each penalizers). So this
+        # should not add hefty computation overhead other than simple checks.
+        #
+        # While we choose not to even create the class instances if they are not required, this
+        # could add additional complexity to the {ScheduleBatch} class, especially we need to
+        # handle {filter_batch()} and {merge()} cases as well.
+        self.penalizer_orchestrator = penaltylib.BatchedPenalizerOrchestrator(
+            vocab_size=vocab_size,
+            batch=self,
+            device=device,
+            Penalizers={
+                penaltylib.BatchedFrequencyPenalizer,
+                penaltylib.BatchedMinNewTokensPenalizer,
+                penaltylib.BatchedPresencePenalizer,
+                penaltylib.BatchedRepetitionPenalizer,
+            },
+        )
+
+        # Handle logit bias but only allocate when needed
+        self.logit_bias = None
+
+    def prepare_for_extend(self, vocab_size: int):
+        bs = self.batch_size()
+        reqs = self.reqs
+        input_ids = [r.fill_ids[len(r.prefix_indices) :] for r in reqs]
+        extend_num_tokens = sum(len(ids) for ids in input_ids)
+        seq_lens = []
+
+        if self.sp_size == 1:
+            flatten_input_ids = sum(input_ids, [])
+        else:
+            flatten_input_ids = seq_parallel_input_ids_extend(
+                input_ids, self.sp_size, bs
+            )
+
+        # Allocate memory
+        req_pool_indices_cpu = self.alloc_req_slots(bs)
+        if self.sp_size > 1:
+            ext_lens = np.asarray(
+                [len(req.fill_ids) - len(req.prefix_indices) for req in reqs]
+            )
+            extend_local_token_nums = seq_parallel_local_len_extend(
+                self.sp_rank, self.sp_size, ext_lens
+            )
+            self.prefill_extend_lens = ext_lens
+            extend_num_tokens = int(np.sum(extend_local_token_nums))
+        out_cache_loc = self.alloc_token_slots(extend_num_tokens)
+
+        pt = 0
+        for i, req in enumerate(reqs):
+            req.req_pool_idx = req_pool_indices_cpu[i]
+            pre_len, seq_len = len(req.prefix_indices), len(req.fill_ids)
+            ext_len = seq_len - pre_len
+            seq_lens.append(seq_len)
+
+            if pre_len > 0:
+                self.req_to_token_pool.req_to_token[req.req_pool_idx][
+                    :pre_len
+                ] = req.prefix_indices
+
+            if self.sp_size > 1:
+                ext_len = extend_local_token_nums[i]
+            self.req_to_token_pool.req_to_token[req.req_pool_idx][pre_len:seq_len] = (
+                out_cache_loc[pt : pt + ext_len]
+            )
```

Comment:
When sp_size > 1, here we only adjusted `ext_len` but did not adjust `pre_len` and `seq_len`, which would cause tensor shape mismatch